transcendence-app/src/shared/casl/casl-ability.factory.ts

```diff
@@ -0,0 +1,40 @@
+import { AbilityBuilder, createMongoAbility } from '@casl/ability';
+import { Injectable } from '@nestjs/common';
+import { UserWithRoles } from '../../user/infrastructure/db/user-with-role.entity';
+import { Action } from '../enums/action.enum';
+import { Role } from '../enums/role.enum';
+import { ChatroomMember } from '../../chat/chatroom/chatroom-member/infrastructure/db/chatroom-member.entity';
+import { Chatroom } from '../../chat/chatroom/infrastructure/db/chatroom.entity';
+
+export class ChatroomWithMembers extends Chatroom {
+  public readonly chatroomMembers: ChatroomMember[];
+
+  constructor(chatroom: Chatroom, chatroomMembers: ChatroomMember[]) {
+    super(chatroom);
+    this.chatroomMembers = [...chatroomMembers];
+  }
+}
+
+@Injectable()
+export class CaslAbilityFactory {
+  defineAbilitiesFor(user: UserWithRoles) {
+    const { can, cannot, build } = new AbilityBuilder(createMongoAbility);
+    can(Action.Manage, 'ChatroomWithMembers', { ownerId: user.id });
+    can(Action.Manage, 'ChatroomWithMembers', {
+      chatroomMembers: { userId: user.id, admin: true },
+    });
+    can(Action.Read, 'ChatroomWithMembers', {
+      chatroomMembers: { userId: user.id },
```

Comment:
`{ userId: user.id }{ userId: user.id }` This is a very non convenient and non idiomatic way to puc conditional logic!

Also, handling the particular field modification cases is hard!